### PR TITLE
Make C++ extensions compile with correct compiler

### PR DESCRIPTION
**Pull request by apmasell**
Some extension files are C++ but have `.c` extensions. GCC will compile these
as C++, but clang will not.

### CODE REVIEW

File: numba/_typeof.cpp [hunk 1]
```diff
@@ -842,6 +842,10 @@ int typecode_devicendarray(PyObject *dispatcher, PyObject *ary)
     int dtype;
     int ndim;
     int layout = 0;
+    PyObject *ndim_obj = nullptr;
+    PyObject* num_obj = nullptr;
+    PyObject* dtype_obj = nullptr;
```
Comment (stuartarchibald) [inline numba/_typeof.cpp:847]:
```suggestion
    PyObject *num_obj = nullptr;
    PyObject *dtype_obj = nullptr;
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @apmasell. I've taken a look through the C/C++ code changes, specifically at things like type casts and the proposed changes seem valid on inspection.

**Review by gmarkall — APPROVED** [review]
This looks good to me - I'm approving it, though I note Stuart's suggestion about the formatting of the pointers on the diff.

**Comment by apmasell** [comment]
Accepted the suggestion. Might be worth extending the formatting from llvmlite to numba.

**Review by stuartarchibald — APPROVED** [review]

**Pull request merged**